### PR TITLE
MSLI-3 - Search

### PR DESCRIPTION
**Pull request by IvanGit2021**
Load screen with List of 100 Most Popular Series ordered by Name (test best results from API);

Use Search View to filter Series from API, ordered by Name;

If no results are found, show TextView with message “No results”;

Each series contains a button to add to Series List or remove if already added (checkmark ☑);

When a Series is clicked, a Series Details screen is opened

### CODE REVIEW

File: My Series List/My Series List/Model/Repository/SeriesRepository.swift [hunk 1]
```diff
@@ -13,18 +13,23 @@ class SeriesRepository {
...
     func getSeries(completionHandler: @escaping (Result<[Series], Error>) -> Void){
-        seriesLocalDataSource.getSeries { coreDataSeries in
-            completionHandler(coreDataSeries)
+        seriesLocalDataSource.getSeries { results in
+            completionHandler(results)
         }
     }
     
-    func deleteSeries(series: Series,  completionHandler: @escaping (Result<String, Error>) -> Void) {
+    func insertSeries(series: Api.Series) {
```
Comment (andre-exd) [inline My Series List/My Series List/Model/Repository/SeriesRepository.swift]: Please consider adding method insertSeries() below existing methods in order to ready changes in historical way

File: My Series List/My Series List/My_Series_List.xcdatamodeld/My_Series_List.xcdatamodel/contents [hunk 2]
```diff
@@ -1,13 +1,14 @@
 <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
 <model type="com.apple.IDECoreDataModeler.DataModel" documentVersion="1.0" lastSavedToolsVersion="19206" systemVersion="20F71" minimumToolsVersion="Automatic" sourceLanguage="Swift" userDefinedModelVersionIdentifier="">
     <entity name="Series" representedClassName=".Series" syncable="YES" codeGenerationType="class">
-        <attribute name="id" optional="YES" attributeType="Integer 32" defaultValueString="0" usesScalarValueType="YES"/>
-        <attribute name="name" optional="YES" attributeType="String"/>
-        <attribute name="overview" optional="YES" attributeType="String"/>
+        <attribute name="id" optional="YES" attributeType="Double" defaultValueString="0" usesScalarValueType="YES"/>
+        <attribute name="isSaved" optional="YES" attributeType="Boolean" usesScalarValueType="YES"/>
```
Comment (andre-exd) [inline My Series List/My Series List/My_Series_List.xcdatamodeld/My_Series_List.xcdatamodel/contents:5]: Is it possible to reorder attributes in order to have same order as in Series from SeriesModel?

File: My Series List/My Series List/View/SearchView/SearchCollectionViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    let searchPresenter = SearchPresenter()
+    var buttonBinding : ((Any) -> Void)?
+    
+    @IBAction func searchCheckMarkPressed(_ sender: UIButton) {
+        if let btnAction = self.buttonBinding {
+            btnAction(self)
+        }
+    }
+}
+
```
Comment (andre-exd) [inline My Series List/My Series List/View/SearchView/SearchCollectionViewCell.swift]: Remove extra lines

File: My Series List/My Series List/View/SearchView/SearchController.swift [hunk 4]
```diff
@@ -30,13 +36,86 @@ extension SearchController: SearchView {
...
+        self.series = series
+        DispatchQueue.main.async {
+            self.collectionView.reloadData()
+        }
     }
     
-    func listEmpty(_ error: Error) {
-        DispatchQueue.main.async {
-            self.listEmptyLabel.text = "Couldn't retreave your data, please try later.\nError: \(error.localizedDescription)"
+    func listError(_ error: Error) {
```
Comment (andre-exd) [inline My Series List/My Series List/View/SearchView/SearchController.swift]: Move listError below existing listEmpty

File: My Series List/My Series List/View/SearchView/SearchController.swift [hunk 5]
```diff
@@ -30,13 +36,86 @@ extension SearchController: SearchView {
...
+        collectionView.reloadData()
+        emptyLabel.isHidden = true
+    }
+    
+    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
+        searchPresenter.searchSeries(search: "")
+        emptyLabel.isHidden = true
+    }
+}
+
```
Comment (andre-exd) [inline My Series List/My Series List/View/SearchView/SearchController.swift]: Remove extra lines

File: My Series List/My Series List/Model/Repository/SeriesRemoteDataSource.swift [hunk 6]
```diff
@@ -9,10 +9,11 @@ import Foundation
 
 class SeriesRemoteDataSource {
     
-    let url = "https://api.themoviedb.org/3/search/tv?api_key=0fbc13d4f4bb3dcc23cb30a1a289d426&language=en-US&page=1&include_adult=false&query="
```
Comment (andre-exd) [inline My Series List/My Series List/Model/Repository/SeriesRemoteDataSource.swift:12]: Can you consider 2 different variables for url?
apiUrl = "https://api.themoviedb.org/3/"
apiParameters = "api_key=fcad3621f48126f0c82cbfb62dc59904"
If you need to define a searchUrl you can concatenate apiUrl with "search/movie?query=" and add apiParameters
Something like this: searchUrl = apiURL + "search/movie?query=" + search + "&" + apiParameters

File: My Series List/My Series List/Model/Repository/SeriesRepository.swift [hunk 7]
```diff
@@ -13,18 +13,23 @@ class SeriesRepository {
...
+        seriesLocalDataSource.insertSeries(series: series) { results in
```
Comment (andre-exd) [inline My Series List/My Series List/Model/Repository/SeriesRepository.swift]: Fix formatting

File: My Series List/My Series List/View/SearchView/SearchPresenter.swift [hunk 8]
```diff
@@ -5,32 +5,84 @@
...
+            checkMark.setImage(UIImage(systemName: "checkmark.rectangle"), for: .normal)
+            seriesSorted[indexPath.row].isSaved = false
+            seriesRepository.getSeries { results in
+                switch results {
+                case .failure(_):
+                    break
+                case .success(let seriesCoreData):
+                    for series in seriesCoreData {
+                        if series.id == self.seriesSorted[indexPath.row].id {
+                            self.seriesRepository.deleteSeries(series: series) { results in
```
Comment (andre-exd) [inline My Series List/My Series List/View/SearchView/SearchPresenter.swift]: Seems like results is not being used. Is there a better way to specify that the parameters are not being used?

File: My Series List/My Series List/View/SearchView/SearchPresenter.swift [hunk 9]
```diff
@@ -5,32 +5,84 @@
...
+        }
+    }
+    
+    func checkRepeated(seriesArray: [Api.Series]) {
+        seriesRepository.getSeries { results in
+            switch results {
+            case .failure(_):
+                break
+            case .success(let seriesCoreData):
+                for seriesA in seriesCoreData {
```
Comment (andre-exd) [inline My Series List/My Series List/View/SearchView/SearchPresenter.swift]: Can you use a better name for distinguishing Series? Given that seriesArray is the parameter of the method, those Series could be called series as it is the main purpose of the method updating some of those

File: My Series List/My Series List/View/SearchView/SearchPresenter.swift [hunk 10]
```diff
@@ -5,32 +5,84 @@
...
+                    for seriesB in seriesArray {
+                        if seriesA.id == seriesB.id! {
```
Comment (andre-exd) [inline My Series List/My Series List/View/SearchView/SearchPresenter.swift]: You should use break in order to skip checking all values.
In my understanding you might even want to switch the order of the for cycles in order to iterate each item from seriesArray once and for each one checking values on repository

**Review by andre-exd — CHANGES_REQUESTED** [review]

**Review by andre-exd — COMMENTED** [review]
Looks good. Please solve conflicts with main

File: My Series List/My Series List/My_Series_List.xcdatamodeld/My_Series_List.xcdatamodel/contents (hunk 2 above)
Comment (IvanGit2021) [reply My Series List/My Series List/My_Series_List.xcdatamodeld/My_Series_List.xcdatamodel/contents:5]: I found that Xcode doesn't allow to change the order of the attributes and it presents it sorted by alphabetic name, would be a possibility to change the Model instead ?

Comment (andre-exd) [reply My Series List/My Series List/My_Series_List.xcdatamodeld/My_Series_List.xcdatamodel/contents:5]: I was not aware of this. Keep it as it is then

**Review by andre-exd — APPROVED** [review]

**Pull request merged**